vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,797 @@
+                     numberOfFails <- numberOfFails + 1
+       } |> Async.Start
+    
+    /// Creates the code lens ui elements for the specified text view line
+    (*
+    let getCodeLensUIElement (lens : CodeLens) (line : ITextViewLine) =
+        let uiContext = SynchronizationContext.Current
+        asyncMaybe {
+            try
+                let! view = view
+                match lens.Computed, isNull lens.UiElement with
+                // The line is already computed and has an existing UI element which is proved to be safe to use
+                | true, false -> 
+                    let textBox = lens.UiElement :?> TextBlock
+                    return Some textBox
+                // The line is already computed but the UI element hasn't been created yet
+                | _, _ ->
+                    let! taggedText, navigation = lens.TaggedText
+                    do! Async.SwitchToContext uiContext |> liftAsync
+                    let textBox = new TextBlock(Width = view.ViewportWidth, Background = Brushes.Transparent, Opacity = 0.5, TextTrimming = TextTrimming.WordEllipsis)
+                    DependencyObjectExtensions.SetDefaultTextProperties(textBox, formatMap.Value)
+                    for text in taggedText do
+                        let run = Documents.Run text.Text
+                        DependencyObjectExtensions.SetTextProperties (run, layoutTagToFormatting text.Tag)
+                        let inl =
+                            match text with
+                            | :? Layout.NavigableTaggedText as nav when navigation.IsTargetValid nav.Range ->
+                                let h = Documents.Hyperlink(run, ToolTip = nav.Range.FileName)
+                                h.Click.Add (fun _ -> navigation.NavigateTo nav.Range)
+                                h :> Documents.Inline
+                            | _ -> run :> _
+                        textBox.Inlines.Add inl
+                    textBox.Opacity <- 0.5
+                    lens.Computed <- true
+                    lens.UiElement <- textBox
+                    if line.IsValid then
+                        let offset = 
+                            view.TextViewLines.GetCharacterBounds(line.Start).Top - view.TextViewLines.GetCharacterBounds(view.TextViewLines.FirstVisibleLine.Start).Top
+                        logInfof "Offset %A" offset
+                        view.DisplayTextLineContainingBufferPosition(line.Start, offset, ViewRelativePosition.Top);
+                    return Some textBox
+                with e -> 
+                    logErrorf "Unexpected exception occured! %A" e
+                    return None
+        } |> Async.map (fun ui ->
+               match ui with
+               | Some (Some ui) ->
+                   Some ui
+               | _ -> 
+                   None)
+    *)
+    do buffer.Changed.AddHandler(fun _ e -> (self.BufferChanged e))
+
+    member __.BufferChanged ___ =
```

Comment:
Does this need to be a separate member?  Also, it is normally better to capture ``self`` via ``member self.BufferChanged __ = ...[self]...`` rather than ``... as self`` where possible.

Reply:
OK, let me have a look whether it must be the case. Likely it's not needed.